powerline/segments/i3wm.py

```diff
@@ -39,27 +48,58 @@ def workspaces(pl, segment_info, only_show=None, output=None, strip=0):
 
 	Highlight groups used: ``workspace`` or ``w_visible``, ``workspace`` or ``w_focused``, ``workspace`` or ``w_urgent``.
 	'''
-	global conn
-	if not conn:
-		try:
-			import i3ipc
-		except ImportError:
-			import i3 as conn
-		else:
-			conn = i3ipc.Connection()
-
 	output = output or segment_info.get('output')
 
 	return [{
-		'contents': w['name'][min(len(w['name']), strip):],
+		'contents': w['name'][strip:],
 		'highlight_groups': calcgrp(w)
-	} for w in conn.get_workspaces()
-		if (not only_show or any(w[typ] for typ in only_show))
-		and (not output or w['output'] == output)
+	} for w in get_i3_connection().get_workspaces()
+		if ((not only_show or any(w[typ] for typ in only_show))
+		    and (not output or w['output'] == output))
```

Comment:
This finally has correct alignment, but apparently this part was not valid from the very beginning: `} for` has the same indentation as `]`, but `{…}` is *deeper* and thus cannot have the same indentation. This should be

```Python
return [
    {
        'contents': w['name'][strip:],
        'highlight_groups': calcgrp(w)
    }
    for w in get_i3_connection().get_workspaces()
    if (
        (not only_show or any((w[typ] for typ in only_show)))
        and (not output or w['output'] == output)
    )
]
```

Replace four spaces with tabs and do not mess with alignment at all.

Though no, I will change this myself just in case.

You can officially call this doomed list comprehension. clint.py from Neovim with my recently merged modifications would have caught this, but it is for C.

---

BTW, I was enclosing generator in another level of parenthesis always since the very beginning, because something was failing with syntax from your code in Python-2.6. But Python documentation shows that this syntax is OK from [Python-2.4](https://docs.python.org/2.6/whatsnew/2.4.html) where it was introduced. Do you have any idea what was failing?